File: scripts/spoof_user_location.ts
import WebSocket, { WebSocketServer } from "ws";

interface Coordinates {
    lat: number;
    long: number;
    accuracy: number;
}

interface UserLocation {
    coords: Coordinates;
    time: number;
}

const wss = new WebSocketServer({ port: 8000 });

let long: -79.3800;
let lat: 43.6550

function rand_drift(val: number, magnitude: 0.00002){
    return val + (Math.random() - 0.5)*magnitude;
}

function broadcast(location: UserLocation){
    const msg = JSON.stringify(location);
    wss.clients.forEach((client) =>{
        if(client.readyState === WebSocket.OPEN) {
            client.send(msg);
        }
    })
}

setInterval(() => {
  lat = rand_drift(lat);
  long = rand_drift(long);

  const payload: UserLocation = {
    coords: {
      lat,
      long,
      accuracy: Math.random() * 10 + 5, // meters
    },
    time: Date.now(),
  };

  broadcast(payload);
  console.log("Sent:", payload);
}, 2000); // every 2 seconds

File: scripts/spoof_user_location.py
import asyncio
import json
import random
import time
import websockets

lat = 43.6550
lon = -79.3800
magnitude = 0.00005

clients = set()

async def broadcast():
    global lat, lon
    while True:
        lat += (random.random() - 0.5) * magnitude
        lon += (random.random() - 0.5) * magnitude

        payload = {
            "coords": {
                "lat": lat,
                "long": lon,
                "accuracy": random.uniform(5, 15)
            },
            "time": int(time.time() * 1000)
        }

        message = json.dumps(payload)

        if clients:
            await asyncio.gather(*[client.send(message) for client in clients])
        print("Sent:", message)
        await asyncio.sleep(2)

async def handler(websocket):
    clients.add(websocket)
    try:
        await websocket.wait_closed()
    finally:
        clients.remove(websocket)

async def main():
    server = await websockets.serve(handler, "0.0.0.0", 8000)
    print("WebSocket server started on ws://localhost:8000")
    await broadcast()

if __name__ == "__main__":
    asyncio.run(main())
